src/jit/morph.cpp

```diff
@@ -18519,6 +18521,280 @@ void Compiler::fgAddFieldSeqForZeroOffset(GenTree* op1, FieldSeqNode* fieldSeq)
     }
 }
 
+class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
+{
+    class Value
+    {
+        GenTree* m_node;
+        bool     m_address;
+        unsigned m_lclNum;
+        ssize_t  m_offset;
+
+    public:
+        Value() : m_node(nullptr), m_address(false), m_lclNum(BAD_VAR_NUM), m_offset(0)
+        {
+        }
+
+        Value(GenTree* node) : m_node(node), m_address(false), m_lclNum(BAD_VAR_NUM), m_offset(0)
+        {
+        }
+
+        GenTree* Node() const
+        {
+            return m_node;
+        }
+
+        bool IsLocation() const
+        {
+            return m_lclNum != BAD_VAR_NUM;
+        }
+
+        unsigned GetLclNum() const
+        {
+            assert(IsLocation());
+
+            return m_lclNum;
+        }
+
+        void Location(unsigned lclNum, unsigned offset = 0)
+        {
+            assert(!IsLocation());
+            assert(!IsAddress());
+
+            m_lclNum = lclNum;
+            m_offset = offset;
+        }
+
+        void Address(const Value& val)
+        {
+            assert(!IsAddress());
+
+            if (val.IsLocation())
+            {
+                m_address = true;
+                m_lclNum  = val.m_lclNum;
+                m_offset  = val.m_offset;
+            }
+        }
+
+        bool IsAddress() const
+        {
+            assert(!m_address || IsLocation());
+
+            return m_address;
+        }
+
+        void Field(const Value& val, unsigned offset)
+        {
+            assert(!IsLocation());
+
+            if (val.IsAddress())
+            {
+                m_lclNum = val.m_lclNum;
+                m_offset = val.m_offset + offset;
```

Comment:
Needs an overflow check. Haven't yet decided what type `m_offset` should be. It's `ssize_t` now because that's what we use in `GenTreeIntCon` but it probably makes more sense to use `unsigned`.

Reply:
Maybe I should just remove the offset for now, the old code didn't handle this anyway.

Reply:
Could you please explain in which case having offset can allow us not to mark lclVars as address taken? I mean in the new code.

Reply:
We need to check both the indir size and the offset to make sure we don't read past the variable, if that happens we need to mark the variable as address exposed. But for now we may be able to just ignore `GT_FIELD` (that is, mark as address exposed anytime we encounter a field, like we do for other unknown nodes). I'll have to run diffs.

Reply:
Yeah, we need to handle `GT_FIELD`, there are CQ regressions without this. And I remembered the exact reason why we need to keep track of the offset - it is possible to have chains of field and address nodes - FIELD(ADDR(FIELD(ADDR(LCLAR)))). You need to keep adding the offsets along the path so when you reach the last field and finally dereference the address you know where you are. It is possible, at least in non-verifiable code, to have ended up outside of the lclvar.

Reply:
>  It is possible, at least in non-verifiable code, to have ended up outside of the lclvar.

I think in this cas ECMA doesn't gurantee anything, does it?

Reply:
Yes, that's certainly undefined behavior. But that IMO applies only to the generated code - it may produce garbage values, corrupt memory, crash etc. However, the compiler itself cannot go down the same path - it should not break its internal invariants (e.g. having address exposed variables that are not marked as such), corrupt memory, crash etc. Perhaps it would be better if the compiler would throw a tomato, err, sorry, InvalidProgramException but obviously we did not do that before and cannot start doing it now.

Though I need to double check what morph does in this case, it may be that it actually produces a LCLFLD with a "bad" offset, so the variable isn't actually address exposed.